compiler/dialects/kernel.py

```diff
@@ -28,36 +36,106 @@ class QuantizedBinaryOp(BinaryOp):
 
 
 @irdl_op_definition
-class MulOp(KernelOp, BinaryOp):
+class MulOp(KernelOp, BinaryOp, Parsable):
     name = "kernel.mul"
     assembly_format = (
         "$lhs `,` $rhs attr-dict `:` type($lhs) `,` type($rhs) `->` type($result)"
     )
 
+    @property
+    def parsing_region(self) -> Region:
+        @Builder.implicit_region(
+            (
+                SSAValue.get(self.lhs).type,
+                SSAValue.get(self.rhs).type,
+                *self.result_types,
+            )
+        )
+        def parsing_region(args: tuple[BlockArgument, ...]) -> None:
+            mul = arith.Muli(args[0], args[1])
+            linalg.YieldOp(mul)
+
+        return parsing_region
+
 
 @irdl_op_definition
-class AddOp(KernelOp, BinaryOp):
+class AddOp(KernelOp, BinaryOp, Parsable):
     name = "kernel.add"
     assembly_format = (
         "$lhs `,` $rhs attr-dict `:` type($lhs) `,` type($rhs) `->` type($result)"
     )
 
+    @property
+    def parsing_region(self) -> Region:
+        @Builder.implicit_region(
+            (
+                SSAValue.get(self.lhs).type,
+                SSAValue.get(self.rhs).type,
+                *self.result_types,
+            )
+        )
+        def parsing_region(args: tuple[BlockArgument, ...]) -> None:
+            add = arith.Addi(args[0], args[1])
+            linalg.YieldOp(add)
+
+        return parsing_region
+
 
 @irdl_op_definition
-class MacOp(KernelOp, BinaryOp):
+class MacOp(KernelOp, BinaryOp, Parsable):
     name = "kernel.mac"
     assembly_format = (
         "$lhs `,` $rhs attr-dict `:` type($lhs) `,` type($rhs) `->` type($result)"
     )
 
+    @property
+    def parsing_region(self) -> Region:
+        @Builder.implicit_region(
+            (
+                SSAValue.get(self.lhs).type,
+                SSAValue.get(self.rhs).type,
+                *self.result_types,
+            )
+        )
+        def parsing_region(args: tuple[BlockArgument, ...]) -> None:
+            mul = arith.Muli(args[0], args[1])
+            mac = arith.Addi(args[2], mul)
+            linalg.YieldOp(mac)
+
+        return parsing_region
+
 
 @irdl_op_definition
-class QMacOp(KernelOp, QuantizedBinaryOp):
+class QMacOp(KernelOp, QuantizedBinaryOp, Parsable):
     name = "kernel.qmac"
     assembly_format = (
         "$lhs `,` $rhs `zp_lhs``:` $zp_lhs `zp_rhs``:` $zp_rhs attr-dict `:`"
         " type($lhs) `,` type($rhs) `,` type($zp_lhs) `,` type($zp_rhs) `->` type($result)"
     )
 
+    @property
+    def parsing_region(self) -> Region:
+        @Builder.implicit_region(
```

Comment:
I find this name a bit confusing, because you are not parsing a region here, you are emitting it to be easier to be parsed, later, right?

Reply:
Can we rename this to `equivalent_region` ?